src/scheduler/Scheduler.cpp

```diff
@@ -224,228 +223,341 @@ faabric::util::SchedulingDecision Scheduler::callFunctions(
         throw std::runtime_error("Message with no master host");
     }
 
-    // Set up scheduling decision
-    SchedulingDecision decision(firstMsg.appid(), firstMsg.groupid());
-
-    // TODO - more granular locking, this is conservative
-    faabric::util::FullLock lock(mx);
-
     // If we're not the master host, we need to forward the request back to the
     // master host. This will only happen if a nested batch execution happens.
-    std::vector<int> localMessageIdxs;
+    SchedulingDecision decision(firstMsg.appid(), firstMsg.groupid());
     if (!forceLocal && masterHost != thisHost) {
-        SPDLOG_DEBUG(
-          "Forwarding {} {} back to master {}", nMessages, funcStr, masterHost);
+        SPDLOG_DEBUG("Forwarding {} back to master {}", funcStr, masterHost);
 
         getFunctionCallClient(masterHost).executeFunctions(req);
         decision.returnHost = masterHost;
         return decision;
     }
 
+    faabric::util::FullLock lock(mx);
+
+    // -------------------------------------------
```

Comment:
Is there any strong argument against abstracting this to a `takeSchedulingDecision` helper method? Then basically scheduling would be a combination of `takeSchedulingDecision` and `doCallFunctions` which seems natural and in-line with that described in the text.

If it is much fuss (e.g. moving a lot of local state) I'm fine with leaving it like this.

Reply:
Yes good point, I think I tried this when writing it originally and like you say it was a bit fiddly, but will have another go.